pceplib/pcep_msg_objects_encoding.c

```diff
@@ -0,0 +1,1483 @@
+ * the TLV header length, but that extra padding must be taken into account by
+ * the enclosing object by rounding up to the next 4 byte boundary.
+ * Example returned lengths:
+ *   normalize_length(4)  =  4, normalize_length(5)  =  8, normalize_length(6)  =  8, normalize_length(7)  =  8,
+ *   normalize_length(8)  =  8  normalize_length(9)  = 12, normalize_length(10) = 12, normalize_length(11) = 12,
+ *   normalize_length(12) = 12, normalize_length(13) = 13...
+ */
+uint16_t normalize_length(uint16_t length)
+{
+    return (length % 4 == 0) ? length : (length + (4 - (length % 4)));
+}
+
+/*
+ * Encoding functions
+ */
+uint16_t pcep_encode_object(struct pcep_object_header* object_hdr, struct pcep_versioning *versioning, uint8_t *buf)
+{
+    initialize_object_coders();
+
+    if (object_hdr->object_class >= MAX_OBJECT_ENCODER_INDEX)
+    {
+        pcep_log(LOG_INFO, "Cannot encode unknown Object class [%d]", object_hdr->object_class);
+        return 0;
+    }
+
+    object_encoder_funcptr obj_encoder = object_encoders[object_hdr->object_class];
+    if (obj_encoder == NULL)
+    {
+        pcep_log(LOG_INFO, "No object encoder found for Object class [%d]", object_hdr->object_class);
+        return 0;
+    }
+
+    uint16_t object_length = OBJECT_HEADER_LENGTH + obj_encoder(object_hdr, versioning, buf + OBJECT_HEADER_LENGTH);
+    double_linked_list_node *node = (object_hdr->tlv_list == NULL ? NULL : object_hdr->tlv_list->head);
+    for (; node != NULL; node = node->next_node)
+    {
+        /* Returns the length of the TLV, including the TLV header */
+        object_length += pcep_encode_tlv((struct pcep_object_tlv_header *) node->data, versioning, buf + object_length);
+    }
+    object_length = normalize_length(object_length);
+    write_object_header(object_hdr, object_length, buf);
+    object_hdr->encoded_object = buf;
+    object_hdr->encoded_object_length = object_length;
+
+    return object_length;
+}
+
+
+/* Object Header
+ *
+ *   0                   1                   2                   3
+ *   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+ *  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *  | Object-Class  |   OT  |Res|P|I|   Object Length (bytes)       |
+ *  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *  |                                                               |
+ *  //                        (Object body)                        //
+ *  |                                                               |
+ *  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *
+ */
+
+void write_object_header(struct pcep_object_header *object_hdr, uint16_t object_length, uint8_t *buf)
+{
+    buf[0] = object_hdr->object_class;
+    buf[1] = ((object_hdr->object_type << 4) |
+              (object_hdr->flag_p == true ? OBJECT_HEADER_FLAG_P : 0x00) |
+              (object_hdr->flag_i == true ? OBJECT_HEADER_FLAG_I : 0x00));
```

Comment:
More of a stylistic thing but generally we just use the boolean value itself rather than add an unnecessary comparison to true/false

Reply:
done